### DISCUSSION TITLE
A simple API to specify an expiration time for individual entries while inserting is needed

### DISCUSSION THREAD

**Discussion by nnsns** [ben-manes/caffeine#1303] [Q&A] [answered]
How is it possible to do something like this?

```
cache.put(key, "value1", 5, TimeUnit.MINUTES);
cache.put(key, "value2", 1, TimeUnit.HOURS);
cache.put(key, "value3", 2, TimeUnit.DAYS);
```

I've seen examples with `expireAfter()` in the documentation, but it looks too complicated for my needs.
As an example of the simplicity of this approach I like this: https://github.com/jhalterman/expiringmap

```
ExpiringMap<String, Connection> map = ExpiringMap.builder()
  .variableExpiration()
  .build();

map.put("connection", connection, ExpirationPolicy.ACCESSED, 5, TimeUnit.MINUTES);
```

How can something like this be done with the current version of Caffeine?

**Comment by ben-manes** [comment, marked as answer]
It is a little tougher in Caffeine because we have to balance more features and higher performance, while also trying to keep the api simple to use. The `expireAfter` configures the cache to support variable expiration and the `policy()` apis gives [feature-specific methods](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Policy.VarExpiration.html) for more advanced usages. The below examples uses a default policy where a new entry never expires if not set on creation and honors the remaining lifetime on every subsequent read or write. The `put(k, v, duration)` then create or updates the entry to set a new expiration time.

```java
Cache<String, Connection> cache = Caffeine.newBuilder()
    .expireAfter(new Expiry<String, Connection>() {
      public long expireAfterCreate(String key, Connection value, long currentTime) {
        return Long.MAX_VALUE;
      }
      public long expireAfterUpdate(String key, Connection connection, 
          long currentTime, long currentDuration) {
        return currentDuration;
      }
      public long expireAfterRead(String key, Connection connection,
          long currentTime, long currentDuration) {
        return currentDuration;
      }
    }).build();
cache.policy().expireVariably().orElseThrow()
    .put("connection", connection, Duration.ofMinutes(5));
```

For more advanced behavior where entries have different policies (some based on time-to-live, others by time-to-idle), you could wrap the value in your own class `Expirable<V>` to let the `Expiry` inspect and act accordingly.
